README.md
# d3-heatmap2

A D3.js plugin that produces heatmaps.

[![Heatmap](heatmap.png)](https://github.com/spiermar/d3-heatmap2/)

## Example

Click [here](http://bl.ocks.org/spiermar/raw/76dc64a7716205df91e51b485a2445e8/) to check the demo, and [source](https://gist.github.com/spiermar/76dc64a7716205df91e51b485a2445e8).

## Installing

If you use NPM, `npm install d3-heatmap2`. Otherwise, download the [latest release](https://github.com/spiermar/d3-heatmap2/releases/latest).

## Getting Started

```html
<head>
  <link rel="stylesheet" type="text/css" href="https://cdn.jsdelivr.net/gh/spiermar/d3-heatmap2@0.1/dist/d3-heatmap2.css">
</head>
<body>
  <div id="chart"></div>
  <div id="details"></div>
  <script src="https://d3js.org/d3.v4.min.js" charset="utf-8"></script>
  <script src="https://cdn.jsdelivr.net/gh/spiermar/d3-heatmap2@0.1/dist/d3-heatmap2.js"></script>
  <script>
    d3.json("data.json", function(error, data) {
        if (error) return console.warn(error)    
        var h = d3.heatmap()
            .width(960)
            .xAxisScale([0,120])
            .yAxisScale([0,2000])
            .colorScale(d3.scaleLinear()
                .domain([0, 65 / 2, 65])
                .range(['#F5F5DC', '#FF5032', '#E50914'])
            )
            
        d3.select("#chart")
            .datum(data)
            .call(h)
    })
  </script>
</body>
```

## Heatmap Data Structure

```json
[
    [1,0,0],
    [0,1,0],
    [0,0,1]
]
```

## API Reference

<a href="#title" name="title">#</a> heatmap.<b>title</b>(<i>[title]</i>)

Heatmap title. Defaults to _empty_ if not set. If called with no arguments, `title` will return the heatmap title. 

<a href="#subtitle" name="subtitle">#</a> heatmap.<b>subtitle</b>(<i>[subtitle]</i>)

Heatmap subtitle. Defaults to _empty_ if not set. If called with no arguments, `subtitle` will return the heatmap subtitle. 

<a href="#legendLabel" name="legendLabel">#</a> heatmap.<b>legendLabel</b>(<i>[label]</i>)

Heatmap label legend. Defaults to _empty_ if not set. If called with no arguments, `legendLabel` will return the heatmap label legend. 

<a href="#width" name="width">#</a> heatmap.<b>width</b>(<i>[width]</i>)

Heatmap width in _px_. Defaults to 960px if not set. If called with no arguments, `width` will return the heatmap width. 

<a href="#margin" name="margin">#</a> heatmap.<b>margin</b>(<i>[object]</i>)

Heatmap title. Defaults to the values below if not set.

```js
{
    top: 20,
    right: 0,
    bottom: 0,
    left: 0
}
```

If called with no arguments, `margin` will return the heatmap margin object. 

<a name="colorScale" href="#colorScale">#</a> heatmap.<b>colorScale</b>(<i>[function]</i>)

Replaces the built-in color scale function. The scale function takes no arguments, and returns a [d3.scale](https://github.com/d3/d3-scale) function. Example:

```js
heatmap.colorScale(d3.scaleLinear()
    .domain([0, 65 / 2, 65])
    .range(['#FFFFFF', '#ff5032', '#e50914'])
);
```

If called with no arguments, `colorScale` will return the heatmap _colorScale_ function.

<a name="xAxisScale" href="#xAxisScale">#</a> heatmap.<b>xAxisScale</b>(<i>[array]</i>)

Heatmap xAxis scale domain, in a value scale array format. Defaults to `null` if not set, which displays the xAxis labels instead of the scale.

```js
heatmap.xAxisScale([0, 1000]);
```

If called with no arguments, `xAxisScale` will return the heatmap xAxis scale domain array.

<a name="yAxisScale" href="#yAxisScale">#</a> heatmap.<b>yAxisScale</b>(<i>[array]</i>)

Heatmap yAxis scale domain, in a value scale array format. Defaults to `null` if not set, which displays the yAxis labels instead of the scale.

```js
heatmap.yAxisScale([0, 100]);
```

If called with no arguments, `yAxisScale` will return the heatmap yAxis scale domain array.

<a name="xAxisLabelFormat" href="#xAxisLabelFormat">#</a> heatmap.<b>xAxisLabelFormat</b>(<i>[function]</i>)

Heatmap xAxis label formatter function. The label formatter function takes a single argument, the label array value, and returns a label string. Defaults to `function (d) { return d }` if not set, which displays the value in the xAxis labels array.

```js
heatmap.xAxisLabelFormat(function (d) { return d + ' sec' });
```

If called with no arguments, `xAxisLabelFormat` will return the heatmap xAxis label formatter function.

<a name="yAxisLabelFormat" href="#yAxisLabelFormat">#</a> heatmap.<b>yAxisLabelFormat</b>(<i>[function]</i>)

Heatmap yAxis label formatter function. The label formatter function takes a single argument, the label array value, and returns a label string. Defaults to `function (d) { return d }` if not set, which displays the value in the yAxis labels array.

```js
heatmap.yAxisLabelFormat(function (d) { return d + ' bucket' });
```

If called with no arguments, `yAxisLabelFormat` will return the heatmap yAxis label formatter function.

<a name="xAxisTickFormat" href="#xAxisTickFormat">#</a> heatmap.<b>xAxisTickFormat</b>(<i>[function]</i>)

Heatmap xAxis tick format function. The tick format function takes a single argument, the tick value, and returns a formatted tick. Defaults to `d3.format('.0f')` if not set.

```js
heatmap.xAxisTickFormat(d3.format('.0f'));
```

If called with no arguments, `xAxisTickFormat` will return the heatmap xAxis tick format function.

<a name="yAxisTickFormat" href="#yAxisTickFormat">#</a> heatmap.<b>yAxisTickFormat</b>(<i>[function]</i>)

Heatmap yAxis tick format function. The tick format function takes a single argument, the tick value, and returns a formatted tick. Defaults to `d3.format('.2s')` if not set.

```js
heatmap.yAxisTickFormat(d3.format('.2s'));
```

If called with no arguments, `yAxisTickFormat` will return the heatmap yAxis tick format function.

<a href="#hideLegend" name="hideLegend">#</a> heatmap.<b>hideLegend</b>(<i>[bool]</i>)

Hide heatmap legend. Defaults to `false` if not set. If called with no arguments, `hideLegend` will return the status of the heatmap legend.

<a name="onClick" href="#onClick">#</a> heatmap.<b>onClick</b>(<i>[function]</i>)

Defines a function that will be executed when the user clicks on a frame. Function takes 3 arguments, the value, the column index and row index respectively.

```js
heatmap.onClick(function (d, i, j) {
    console.info("Clicked on row " + j + ", column " + i + ", value " + d);
});
```

If called with no arguments, `onClick` will return the click handler.

<a name="onMouseOver" href="#onMouseOver">#</a> heatmap.<b>onMouseOver</b>(<i>[function]</i>)

Defines a function that will be executed when the user hovers the mouse over a frame. Function takes 3 arguments, the value, the column index and row index respectively.

```js
heatmap.onMouseOver(function (d, i, j) {
    document.getElementById("details").innerHTML = "column: " + i + ", row: " +  j + ", value: " + d
});
```

If called with no arguments, `onMouseOver` will return the click handler.

<a name="xAxisLabels" href="#xAxisLabels">#</a> heatmap.<b>xAxisLabels</b>(<i>[array]</i>)

Heatmap xAxis labels array. Defaults to `null` if not set, which hides de xAxis label.

```js
heatmap.xAxisLabels([0, 1, 2, ...]);
```

If called with no arguments, `xAxisLabels` will return the heatmap xAxis labels array.

<a name="yAxisLabels" href="#yAxisLabels">#</a> heatmap.<b>yAxisLabels</b>(<i>[array]</i>)

Heatmap yAxis labels array. Defaults to `null` if not set, which hides de yAxis label.

```js
heatmap.yAxisLabels([0, 1, 2, ...]);
```

If called with no arguments, `yAxisLabels` will return the heatmap yAxis labels array.

<a name="highlightColor" href="#highlightColor">#</a> heatmap.<b>highlightColor</b>(<i>[string]</i>)

Heatmap highlight base color. Defaults to `#936EB5` if not set. If called with no arguments, `highlightColor` will return the heatmap highlight base color.

<a name="highlightOpacity" href="#highlightOpacity">#</a> heatmap.<b>highlightOpacity</b>(<i>[string]</i>)

Heatmap highlight opacity. Defaults to `0.4` if not set. If called with no arguments, `highlightOpacity` will return the heatmap highlight opacity.

<a name="setHighlight" href="#setHighlight">#</a> heatmap.<b>setHighlight</b>(<i>[string]</i>)

Set heatmap highlight ranges. Defaults to `[]` if not set, which doesn't highlight any frame. 

```js
heatmap.setHighlight(
    [
        {"start": [0,0], "end": [0,3]},
        {"start": [1,0], "end": [1,3]}
    ]
);
```

The highlight range object must contain a start property and an end property. Start and end properties are tuples, `[column, row]` coordinates, in a two integer array format.

If called with no arguments, `setHighlight` will return the selected heatmap highlight ranges.

<a name="updateHighlight" href="#updateHighlight">#</a> heatmap.<b>updateHighlight</b>()

Updates the heatmap highlights based on the ranges set by the `setHighlight` function. `updateHighlight` is executed automatically at first draw, if highlight ranges are set.

```js
heatmap.updateHighlight();
```

`updateHighlight` takes no arguments and returns `void`.

## Acknowledgements

Thank you [Nadieh Bremer](https://twitter.com/NadiehBremer) for the visual inspiration and [D3.js examples](https://www.visualcinnamon.com/2016/05/smooth-color-legend-d3-svg-gradient.html).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased](https://github.com/spiermar/d3-heatmap2/compare/0.2.0...master)

## [0.2.0](https://github.com/spiermar/d3-heatmap2/compare/0.1.0...0.2.0) - 2018-02-27

- Highlight cells
- Select range example with `hover` option

src/heatmap.js
/* global d3 */

function cantorPair (x, y) {
  var z = ((x + y) * (x + y + 1)) / 2 + y
  return z
}

export default function () {
  var svg = null
  var columns = 0
  var rows = 0
  var title = ''
  var subtitle = ''
  var legendLabel = ''
  var width = 960
  var margin = {
    top: 20,
    right: 0,
    bottom: 0,
    left: 0
  }
  var gridSize = null
  var colorScale = null

  var xAxisScale = null
  var yAxisScale = null
  var xAxisTickFormat = d3.format('.0f')
  var yAxisTickFormat = d3.format('.2s')

  var xAxisLabels = null
  var yAxisLabels = null
  var xAxisLabelFormat = function (d) { return d }
  var yAxisLabelFormat = function (d) { return d }

  var hideLegend = false

  var clickHandler = null
  var mouseOverHandler = null

  var highlight = []
  var highlightColor = '#936EB5'
  var highlightOpacity = '0.4'

  function click (d, i, j) {
    if (typeof clickHandler === 'function') {
      clickHandler(d, i, j)
    }
  }

  function mouseOver (d, i, j) {
    if (typeof mouseOverHandler === 'function') {
      mouseOverHandler(d, i, j)
    }
  }

  function getHighlightFrames () {
    var highlightFrames = []
    for (var k in highlight) { // all highlights
      if (highlight[k].start[0] <= highlight[k].end[0]) { // no reverse column range highlight
        for (var i = highlight[k].start[0]; i <= highlight[k].end[0]; i++) {
          var j = null
          if (i > highlight[k].start[0] && i < highlight[k].end[0]) { // middle columns
            for (j = 0; j < rows; j++) {
              highlightFrames.push([i, j])
            }
          } else if (i === highlight[k].start[0]) { // start column, or start and end are the same
            if (i === highlight[k].end[0]) { // ends in the same column
              if (highlight[k].start[1] <= highlight[k].end[1]) { // no reverse row range highlight
                for (j = highlight[k].start[1]; j <= highlight[k].end[1]; j++) {
                  highlightFrames.push([i, j])
                }
              } else {
                console.log('Error: Start row is higher than end row. No reverse range highlight.')
              }
            } else { // doesn't end in the same column
              for (j = highlight[k].start[1]; j < rows; j++) {
                highlightFrames.push([i, j])
              }
            }
          } else { // end column, when different than start column
            for (j = highlight[k].end[1]; j >= 0; j--) {
              highlightFrames.push([i, j])
            }
          }
        }
      } else {
        console.log('Error: Start column is higher than end column. No reverse range highlight.')
      }
    }
    return highlightFrames
  }

  function updateHighlight () {
    if (highlight && highlight.length > 0 && svg && rows && gridSize) {
      var highlightFrames = getHighlightFrames()
      var frames = svg.selectAll('g.highlight')
        .data(highlightFrames, function (d) { return cantorPair(d[0], d[1]) })
      frames.exit().remove()
      frames.enter().append('g')
        .attr('class', 'highlight')
        .append('rect')
        .attr('x', function (d) { return d[0] * gridSize })
        .attr('y', function (d) { return d[1] * gridSize })
        .attr('width', gridSize)
        .attr('height', gridSize)
        .style('fill', highlightColor)
        .style('fill-opacity', highlightOpacity)
        .style('pointer-events', 'none')
    } else {
      console.log("Error: Can't update highlight. Heatmap was not initialized yet or highlight was not defined.")
    }
  }

  function heatmap (selection) {
    var data = selection.datum()

    columns = data.length
    rows = data[0].length

    gridSize = width / columns
    var height = gridSize * (rows + 2)

    if (title) {
      margin.top = margin.top + 50
    }

    if (subtitle) {
      margin.top = margin.top + 20
    }

    if (!hideLegend) {
      margin.bottom = margin.bottom + 50
    }

    if (yAxisScale || yAxisLabels) {
      margin.left = margin.left + 50
    }

    var max = 0
    for (let i = 0; i < data.length; i++) {
      for (let j = 0; j < data[i].length; j++) {
        if (data[i][j] > max) { max = data[i][j] }
      }
    }

    if (!colorScale) {
      colorScale = d3.scaleLinear()
        .domain([0, max / 2, max])
        .range(['#FFFFDD', '#3E9583', '#1F2D86'])
        // .interpolate(d3.interpolateHcl);
    }

    svg = selection
      .append('svg')
      .attr('width', width + margin.left + margin.right)
      .attr('height', height + margin.top + margin.bottom)
      .append('g')
      .attr('transform', 'translate(' + margin.left + ',' + margin.top + ')')

    if (yAxisScale || yAxisLabels) {
      if (yAxisScale) {
        var y = d3.scaleLinear()
          .domain(yAxisScale)
          .range([height, 0])

        svg.append('g')
          .attr('transform', 'translate(3,-12)')
          .attr('class', 'rowLabel axis')
          .call(d3.axisLeft(y)
            .ticks(20)
            .tickFormat(yAxisTickFormat))
      } else {
        svg.selectAll('.rowLabel')
          .data(yAxisLabels)
          .enter().append('text')
          .text(yAxisLabelFormat)
          .attr('x', 0)
          .attr('y', function (d, i) { return i * gridSize })
          .style('text-anchor', 'end')
          .attr('transform', 'translate(-6,' + gridSize / 1.2 + ')')
          .attr('class', 'rowLabel mono axis')
      }
    }

    if (xAxisScale || xAxisLabels) {
      if (xAxisScale) {
        var x = d3.scaleLinear()
          .domain(xAxisScale)
          // .range([0, width - margin.left - margin.right - 40])
          .range([0, width - margin.left - margin.right])

        svg.append('g')
          .attr('transform', 'translate(5,3)')
          .attr('class', 'columnLabel axis')
          .call(d3.axisTop(x)
            .ticks(20)
            .tickFormat(xAxisTickFormat))
      } else {
        svg.selectAll('.columnLabel')
          .data(xAxisLabels)
          .enter().append('text')
          .text(xAxisLabelFormat)
          .attr('y', function (d, i) { return i * gridSize })
          .attr('x', 0)
          .style('text-anchor', 'beginning')
          .attr('transform', 'translate(' + gridSize / 1.4 + ', -6) rotate(270)')
          .attr('class', 'columnLabel mono axis')
      }
    }

    svg.selectAll('g.column')
      .data(data)
      .enter().append('g')
      .each(function (d, i) { // function (d, i, j) might replace .each.
        d3.select(this).selectAll('rect')
          .data(d)
          .enter().append('rect')
          .attr('x', function (d) { return i * gridSize }) // column
          .attr('y', function (d, j) { return j * gridSize }) // row
          .attr('class', 'bordered')
          .attr('width', gridSize)
          .attr('height', gridSize)
          .style('stroke', 'white')
          .style('stroke-opacity', 0.6)
          .style('fill', function (d) { return colorScale(d) })
          .style('pointer-events', 'all')
          .on('mouseover', function (d, j) { return mouseOver(d, i, j) })
          .on('click', function (d, j) { return click(d, i, j) })
      })

    if (highlight && highlight.length > 0) {
      updateHighlight()
    }

    // Append title to the top
    if (title) {
      svg.append('text')
        .attr('class', 'title')
        .attr('x', width / 2)
        .attr('y', -60)
        .style('text-anchor', 'middle')
        .text(title)
    }

    if (subtitle) {
      svg.append('text')
        .attr('class', 'subtitle')
        .attr('x', width / 2)
        .attr('y', -40)
        .style('text-anchor', 'middle')
        .text(subtitle)
    }

    if (!hideLegend) {
      // Extra scale since the color scale is interpolated
      var countScale = d3.scaleLinear()
        .domain([0, max])
        .range([0, width])

      // Calculate the variables for the temp gradient
      var numStops = 10
      var countRange = countScale.domain()
      var countPoint = []

      countRange[2] = countRange[1] - countRange[0]
      for (var i = 0; i < numStops; i++) {
        countPoint.push(i * countRange[2] / (numStops - 1) + countRange[0])
      }// for i

      // Create the gradient
      svg.append('defs')
        .append('linearGradient')
        .attr('id', 'legend-traffic')
        .attr('x1', '0%').attr('y1', '0%')
        .attr('x2', '100%').attr('y2', '0%')
        .selectAll('stop')
        .data(d3.range(numStops))
        .enter().append('stop')
        .attr('offset', function (d, i) {
          return countScale(countPoint[i]) / width
        })
        .attr('stop-color', function (d, i) {
          return colorScale(countPoint[i])
        })

      var legendWidth = Math.min(width * 0.8, 400)
      // Color Legend container
      var legendsvg = svg.append('g')
        .attr('class', 'legendWrapper')
        .attr('transform', 'translate(' + (width / 2) + ',' + (gridSize * rows + 40) + ')')

      // Draw the Rectangle
      legendsvg.append('rect')
        .attr('class', 'legendRect')
        .attr('x', -legendWidth / 2)
        .attr('y', 0)
        // .attr("rx", hexRadius*1.25/2)
        .attr('width', legendWidth)
        .attr('height', 10)
        .style('fill', 'url(#legend-traffic)')

      // Append title
      legendsvg.append('text')
        .attr('class', 'legendTitle')
        .attr('x', 0)
        .attr('y', -10)
        .style('text-anchor', 'middle')
        .text(legendLabel)

      // Set scale for x-axis
      var xScale = d3.scaleLinear()
        .range([-legendWidth / 2, legendWidth / 2])
        .domain([0, max])

      // Define x-axis
      var xAxis = d3.axisBottom()
        .ticks(5)
        // .tickFormat(formatPercent)
        .scale(xScale)

      // Set up X axis
      legendsvg.append('g')
        .attr('class', 'axis')
        .attr('transform', 'translate(0,' + (10) + ')')
        .call(xAxis)
    }
  }

  heatmap.title = function (_) {
    if (!arguments.length) { return title }
    title = _
    return heatmap
  }

  heatmap.subtitle = function (_) {
    if (!arguments.length) { return subtitle }
    subtitle = _
    return heatmap
  }

  heatmap.legendLabel = function (_) {
    if (!arguments.length) { return legendLabel }
    legendLabel = _
    return heatmap
  }

  heatmap.width = function (_) {
    if (!arguments.length) { return width }
    width = _
    return heatmap
  }

  heatmap.margin = function (_) {
    if (!arguments.length) { return margin }
    margin = _
    return heatmap
  }

  heatmap.colorScale = function (_) {
    if (!arguments.length) { return colorScale }
    colorScale = _
    return heatmap
  }

  heatmap.xAxisScale = function (_) {
    if (!arguments.length) { return xAxisScale }
    xAxisScale = _
    return heatmap
  }

  heatmap.yAxisScale = function (_) {
    if (!arguments.length) { return yAxisScale }
    yAxisScale = _
    return heatmap
  }

  heatmap.xAxisLabelFormat = function (_) {
    if (!arguments.length) { return xAxisLabelFormat }
    xAxisLabelFormat = _
    return heatmap
  }

  heatmap.yAxisLabelFormat = function (_) {
    if (!arguments.length) { return yAxisLabelFormat }
    yAxisLabelFormat = _
    return heatmap
  }

  heatmap.xAxisTickFormat = function (_) {
    if (!arguments.length) { return xAxisTickFormat }
    xAxisTickFormat = _
    return heatmap
  }

  heatmap.yAxisTickFormat = function (_) {
    if (!arguments.length) { return yAxisTickFormat }
    yAxisTickFormat = _
    return heatmap
  }

  heatmap.hideLegend = function (_) {
    if (!arguments.length) { return hideLegend }
    hideLegend = _
    return heatmap
  }

  heatmap.onClick = function (_) {
    if (!arguments.length) { return clickHandler }
    clickHandler = _
    return heatmap
  }

  heatmap.onMouseOver = function (_) {
    if (!arguments.length) { return mouseOverHandler }
    mouseOverHandler = _
    return heatmap
  }

  heatmap.xAxisLabels = function (_) {
    if (!arguments.length) { return xAxisLabels }
    xAxisLabels = _
    return heatmap
  }

  heatmap.yAxisLabels = function (_) {
    if (!arguments.length) { return yAxisLabels }
    yAxisLabels = _
    return heatmap
  }

  heatmap.highlightColor = function (_) {
    if (!arguments.length) { return highlightColor }
    highlightColor = _
    return heatmap
  }

  heatmap.highlightOpacity = function (_) {
    if (!arguments.length) { return highlightOpacity }
    highlightOpacity = _
    return heatmap
  }

  heatmap.setHighlight = function (_) {
    if (!arguments.length) { return highlight }
    highlight = _
    return heatmap
  }

  heatmap.updateHighlight = updateHighlight

  return heatmap
}
